lib/src/main/java/me/onebone/toolbar/CollapsingToolbarScaffold.kt

```diff
@@ -45,6 +49,78 @@ class CollapsingToolbarScaffoldState(
 		get() = offsetYState.value
 
 	internal val offsetYState = mutableStateOf(initialOffsetY)
+
+	// TODO: Maybe should move toolbar expand/collapse methods to CollapsingToolbarState
+	//  but offset expand/collapse leave in CollapsingToolbarScaffoldState?
+
+	// TODO: A strange jump in snap speed is often observed
+	@ExperimentalToolbarApi
+	suspend fun expand(duration: Int = CollapsingToolbarDefaults.ExpandDuration) {
+		val anim = AnimationState(toolbarState.height.toFloat())
+
+		toolbarState.scroll {
+			var prev = anim.value
+			anim.animateTo(toolbarState.maxHeight.toFloat(), tween(duration)) {
+				scrollBy(value - prev)
+				prev = value
+			}
+		}
+	}
+
+	// TODO: A strange jump in snap speed is often observed
+	@ExperimentalToolbarApi
+	suspend fun collapse(duration: Int = CollapsingToolbarDefaults.CollapseDuration) {
+		val anim = AnimationState(toolbarState.height.toFloat())
+
+		toolbarState.scroll {
+			var prev = anim.value
+			anim.animateTo(toolbarState.minHeight.toFloat(), tween(duration)) {
+				scrollBy(value - prev)
+				prev = value
+			}
+		}
+	}
+
+	@ExperimentalToolbarApi
+	suspend fun expandOffset(snapStrategy: SnapStrategy) {
+		val anim = AnimationState(offsetYState.value.toFloat())
+
+		anim.animateTo(0f, tween(snapStrategy.expandDuration)) {
+			offsetYState.value = value.toInt()
+		}
+	}
+
+	@ExperimentalToolbarApi
+	suspend fun collapseOffset(snapStrategy: SnapStrategy) {
+		val anim = AnimationState(offsetYState.value.toFloat())
+
+		anim.animateTo(-toolbarState.minHeight.toFloat(), tween(snapStrategy.collapseDuration)) {
+			offsetYState.value = value.toInt()
+		}
+	}
+
+	// TODO: Is there a better solution rather OptIn ExperimentalToolbarApi?
+	@OptIn(ExperimentalToolbarApi::class)
+	internal suspend fun processSnap(strategy: SnapStrategy) {
+		if (toolbarState.progress > strategy.edge) {
+			expand(strategy.expandDuration)
+		} else {
+			collapse(strategy.collapseDuration)
+		}
+	}
```

Comment:
I see no reason for `processSnap` to reside in ScaffoldState. You can put it as an extension function like this:

```kotlin
@OptIn(ExperimentalToolbarApi::class)
internal suspend fun SnapStrategy.performTo(state: CollapsingToolbarState) {
	if (state.progress > edge) {
		state.expand(expandDuration)
	} else {
		state.collapse(collapseDuration)
	}
}
```

Reply:
It makes sense if the context parameter is `CollapsingToolbarState` as well.

```kotlin
@OptIn(ExperimentalToolbarApi::class)
internal suspend fun CollapsingToolbarState.performSnap(strategy: SnapStrategy) {
	if (state.progress > edge) {
		expand(strategy.expandDuration)
	} else {
		collapse(strategy.collapseDuration)
	}
}
```

Reply:
Sounds good but I need to research would extensions be slower than regular method. Also I don't know where to place this extensions. I think they should be private and store inside `SnapStrategy` file.